### PR TITLE
DaVinci resolves trash effects correctly

### LINKED ISSUE #4987: Davinci + Draw on Trash Effects
At the moment DaVinci is not quite properly implemented. Effects that happen when a card is trashed (like Geist's draw) do not trigger until after you choose a target. While there needs to be a valid target in your hand to use DaVinci's ability the card you install does not need to be in your hand when you trash it. 

This is the same behaviour that is currently properly implemented on Self Modifying code. You draw a card before searching the stack.

### PR DESCRIPTION
**Pull request by Kubik161**
Closes #4987 

Fix and new test.

### CODE REVIEW

File: test/clj/game/cards/programs_test.clj [hunk 1]
```diff
@@ -1438,6 +1438,22 @@
...
+      (new-game {:runner {:id "Armand \"Geist\" Walker: Tech Lord"
+                          :deck ["Simulchip"]
+                          :hand ["DaVinci" "The Turning Wheel"]}})
+      (take-credits state :corp)
+      (play-from-hand state :runner "DaVinci")
+      (let [davinci (get-program state 0)]
+        (core/add-counter state :runner davinci :power 2)
+        (changes-val-macro
+          0 (:credit (get-runner))
+          "DaVinci installs The Turning Wheel for free"
```
Comment (NoahTheDuke) [inline test/clj/game/cards/programs_test.clj:1453]: Should this be Simulchip?

File: test/clj/game/cards/programs_test.clj [hunk 2]
```diff
@@ -1438,6 +1438,22 @@
           (card-ability state :runner (refresh davinci) 0)
           (click-card state :runner "The Turning Wheel"))
         (is (get-resource state 0) "The Turning Wheel is installed")
+        (is (find-card "DaVinci" (:discard (get-runner))) "DaVinci is trashed"))))
+  (testing "Using ability should trigger trash effects first. Issue #4987"
+    (do-game
+      (new-game {:runner {:id "Armand \"Geist\" Walker: Tech Lord"
+                          :deck ["Simulchip"]
+                          :hand ["DaVinci" "The Turning Wheel"]}})
```
Comment (NoahTheDuke) [inline test/clj/game/cards/programs_test.clj:1446]: Is The Turning Wheel needed in this test?

File: test/clj/game/cards/programs_test.clj (hunk 1 above)
Comment (Kubik161) [reply test/clj/game/cards/programs_test.clj]: y

File: test/clj/game/cards/programs_test.clj (hunk 2 above)
Comment (Kubik161) [reply test/clj/game/cards/programs_test.clj:1446]: It's need. You have to have some valid target for DaVinci in hand. But you can then install new card (Simulchip) that has been drawn from deck via Geist ability.

Comment (NoahTheDuke) [reply test/clj/game/cards/programs_test.clj:1446]: Cool, thank you

**Pull request merged**